src/object/srv_ec_aggregate.c

```diff
@@ -560,224 +803,1093 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+		rc = agg_process_holes(entry);
+	else if (update_vos && rc == 0) {
+		if (rc == 0 && entry->ae_oca->u.ec.e_p > 1)  {
+			/* offload of ds_obj_update to push remote parity */
+			rc = agg_peer_update(entry, write_parity);
+			if (rc)
+				D_ERROR("agg_peer_update fail: "DF_RC"\n",
+					DP_RC(rc));
+		}
+
+		if (rc == 0) {
+			rc = agg_update_vos(entry, write_parity);
+			if (rc)
+				D_ERROR("agg_update_vos failed: "DF_RC"\n",
+					DP_RC(rc));
+		}
+	}
+	if (rc == 0 && entry->ae_cur_stripe.as_ho_ext_cnt) {
+		if (entry->ae_oca->u.ec.e_p > 1 && !update_vos)  {
+			/* No VOS update, but holdovers.
+			 * Didn't call peer update,
+			 * so send holdover list to peer for
+			 * removal
+			 */
+			rc = agg_peer_update(entry, false);
+			if (rc)
+				D_ERROR("agg_peer_update failed: "DF_RC"\n",
+					DP_RC(rc));
+		}
+		if (rc == 0) {
+			rc = agg_remove_holdovers(entry);
+			if (rc)
+				D_ERROR("remove_holdovers failed: "DF_RC"\n",
+					DP_RC(rc));
+		}
+	}
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* Returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	daos_off_t		 cur_stripenum, this_stripenum;
+	int			 rc = 0;
+
+	D_ASSERT(!(entry->ie_recx.rx_idx & PARITY_INDICATOR));
+
+	this_stripenum = agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+
+	if (this_stripenum != agg_entry->ae_cur_stripe.as_stripenum) {
+		/* Iterator has reached next stripe */
+		if (agg_entry->ae_cur_stripe.as_extent_cnt) {
+			cur_stripenum = agg_entry->ae_cur_stripe.as_stripenum;
+			rc = agg_process_stripe(agg_entry);
+			if (rc)
+				D_ERROR("Process stripe returned "DF_RC"\n",
+					DP_RC(rc));
+		/* Error leaves data covered by replicas vulnerable to vos
+		 * delete, so don't advance coordination epoch.
+		 */
+			rc = 0;
+			if (cur_stripenum <
+			    agg_entry->ae_cur_stripe.as_stripenum &&
+			agg_entry->ae_cur_stripe.as_stripenum <
+			this_stripenum) {
+				/* Handle holdover stripe */
+				rc = agg_process_stripe(agg_entry);
+				if (rc)
+					D_ERROR("Holdover returned "DF_RC"\n",
+						DP_RC(rc));
+				rc = 0;
+			}
+		}
+		agg_entry->ae_cur_stripe.as_stripenum = this_stripenum;
+	}
+
+	/* Add the extent to the entry, for the current stripe */
+	D_ALLOC_PTR(extent);
+	if (extent == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	if (entry->ie_csum.cs_type)
+		return 1;
```

Comment:
should this move to the beginning of this function? at least extent is not free here?

Reply:
This is just a placeholder for when checksum support is added.  For now, every extent gets looked at, but all are skipped, if the container supports CSUM.  The extents are still cleaned up at the end of each stripe.

Reply:
But it does not link the extent to as_dextents yet?

Reply:
I moved the check/return to above the allocation of the extent, so this should fix the leak.  Once csum support is added this can all go away.  Thanks.